packages/processing/pitop/processing/algorithms/ball_detect.py

```diff
@@ -162,181 +247,92 @@ def __init__(self, image_processing_width: int = 320, format: str = "OpenCV"):
         :param str format: output image format
         """
         import_libs()
-
-        self._image_processing_width = image_processing_width
+        self.image_processing_width = image_processing_width
         self.format = format
-        self.balls = {c: Ball(c) for c in VALID_COLORS}
-        self._frame_scaler = None
+        self.detectors = {}
 
         # Enable FPS if environment variable is set
         self._print_fps = getenv("PT_ENABLE_FPS", "0") == "1"
         if self._print_fps:
             self._fps = FPS().start()
-            atexit.register(self.__print_fps)
-
-    def __call__(self, frame, color: Union[str, list] = "red"):
-        def parse_colors(color_arg):
-            colors = []
-            if isinstance(color_arg, str):
-                if color_arg not in VALID_COLORS:
-                    raise ValueError(
-                        f"Valid color values are {', '.join(VALID_COLORS[:-1])} or {VALID_COLORS[-1]}"
-                    )
-                colors = [color_arg]
-            elif type(color_arg) in (list, tuple):
-                if not set(color_arg).issubset(VALID_COLORS):
-                    raise ValueError(
-                        f"Valid color values are {', '.join(VALID_COLORS[:-1])} or {VALID_COLORS[-1]}"
-                    )
-                colors = color_arg
-            if len(colors) > 3:
-                raise ValueError("Cannot pass more than three colors.")
-
-            return colors
+            atexit.register(self._print_fps)
 
-        import_libs()
-        frame = ImageFunctions.convert(frame, format="OpenCV")
-
-        if self._frame_scaler is None:
-            _, width = frame.shape[0:2]
-            self._frame_scaler = width / self._image_processing_width
+    @property
+    def balls(self):
+        return {c: self.detectors[c].ball for c in self.detectors}
 
-        for c in parse_colors(color):
-            self.balls[c] = self.__find_most_likely_ball(
-                ball=self.balls.get(c), frame=frame, color=c
-            )
+    def add_color_hsv(self, color: str, lower: tuple, upper: tuple):
```

Comment:
I think this should be called add_hsv_color as that is what is called below. I think a better name might be add_color, and then to support various formats use named args e.g. `add_color('yellowish', hsv={ "lower": (...), "upper": (...)})`